src/app/games/martian-chess/MartianChessDummyMinimax.ts

```diff
@@ -0,0 +1,171 @@
+import { Coord } from 'src/app/jscaip/Coord';
+import { Direction } from 'src/app/jscaip/Direction';
+import { Minimax } from 'src/app/jscaip/Minimax';
+import { NodeUnheritance } from 'src/app/jscaip/NodeUnheritance';
+import { Player } from 'src/app/jscaip/Player';
+import { GameStatus } from 'src/app/jscaip/Rules';
+import { MGPOptional } from 'src/app/utils/MGPOptional';
+import { MartianChessMove } from './MartianChessMove';
+import { MartianChessMoveResult, MartianChessNode, MartianChessRules } from './MartianChessRules';
+import { MartianChessPiece, MartianChessState } from './MartianChessState';
+
+export class MartianChessDummyMinimax extends Minimax<MartianChessMove, MartianChessState, MartianChessMoveResult> {
```

Comment:
commentaire général: j'ai l'impression qu'il y a pas mal de logique dupliquée entre Rules et Minimax, ça serait top d'arriver à avoir la majorité dans Rules, et que Minimax n'implémente pas de logique lui-même, outre s'assurer de faire tous les mouvements

Reply:
Tu veux que pour vérifier si un mouvement est légal les Rules fassent leurs liste elles même et check "is in legal list" ?
Si ça diminue la duplication pourquoi pas mais sinon, c'est p-e un peu mochet ?

Reply:
pas tellement qu'elles fassent leur liste, mais que la plupart des check puissent être faits dans les rules, le minimax se limitant à lister les mouvements potentiellement possibles, puis à les checker par les rules. Après j'imagine il y a toujours un certain niveau de duplication, mais jusqu'ici la plupart des jeux minimisent bien cela il me semble

Reply:
Les minimax doivent aussi calculer eux même la logique de comment ils calculent en fonctions des métriques qui les intéressent, mais je sais pas à quel point (prenant l'example des Tafl) getMinimalNumberOfMoveNeededByKingToReachACorner a sa place dans Rules plus que dans un Minimax, si ?

Reply:
Mais oui, luttons contre la duplication. J'avais eu un projet ou un solar-qube était déjà en place, c'était sympa il te listait les duplications enlevables, toi qu'est le délégué devhops si tu te sent de faire ça :p

Reply:
ah oui j'avais trouvé un outil de détection de duplication, on en a moulte à enlever t'inquiètes :p
Mais du coup pas de tentative de simplifier le minimax ici ?

Reply:
Bah, là dans l'état du code le meilleurs moyen de minimiser le minimax c'est de simplifier le rules en disant que isLegal fait
1. Rules.getListMoves
2. return "isAmongstMoveList(butThenWeDon'tReallyGetTheReasonForTheMoveIllegality)"
qui du coup me semble peu souhaitable.

Quel est-il ce fameux outil ?

Reply:
Oui non effectivement ça c'est la contre-solution :D Mais bref, laissons ainsi du coup.
jscpd ou du genre, je t'invite à googler « javascript duplication tool »

Reply:
Bah écoute euh, ouais j'suis auch !

Reply:
J'ai lancé: `npx jscpd src/app` et il me dit: 895 clones trouvés, 13% du code est cloné D:
Ça se base sur des clones de 5 lignes, j'ai un peu varié pour trouver les plus gros:

  - 5 lignes: 895 clones
  - 10 lignes: 304 clones
  - 20 lignes: 31 clones
  - 30 lignes: 3 clones

À voir comment intégrer ça dans notre workflow. C'est un peu comme le coup des fonctions trop longues, je dirais partir des 30 lignes, puis aller vers 20, et continuer à diminuer petit à petit. Sauf que faut trouver une façon de vraiment y penser :D Peut-être à intégrer au ticket multi-sprint? Améliorer 1 jeu + réduire une fonction complexe + enlever un clone ?

Reply:
Mais tout à fait, no goal hygiènes dans l'ordre dans les sprintly ticket, on avait généraliser le GWT, RulesUtils, 100% de coverage, on peut ajouter un différent qui est "Faire un TODO", refactor un de la liste de refactor, enlever une fonction trop longue ou un warning, et enlever des clones, ça fait une belle liste et des tickets inutiles business mais qui font BIEN PLAIZ (c'pas des ticket à burn out ça, pour moi, c'est sûr :D )

Reply:
Ok hagrid donc ! On formalise comment ce process ? Tu rajoute dans la description du ticket pour la v2 de ce ticke ?

Reply:
J'ai ajouté "enlever code-duplicate" dans le enhance-sprintly (bon ce sera que pour les jeux là)
Sinon tu peux relancer (dès que j'aurai poussé) ?
Et me dire ce que tu penses du duplicate sur les .svg et d'à quel point refactor ça aiderait ?